services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusControllerITest.kt

```diff
@@ -200,4 +213,302 @@ class HakemusControllerITest(@Autowired override val mockMvc: MockMvc) : Control
+            val request =
+                HakemusUpdateRequestFactory.createFilledJohtoselvityshakemusUpdateRequest()
+            every {
+                applicationAuthorizer.authorizeApplicationId(
+                    id,
+                    PermissionCode.EDIT_APPLICATIONS.name
+                )
+            } returns true
+            every { hakemusService.updateHakemus(id, request, USERNAME) } throws
+                InvalidHakemusyhteystietoException(
+                    id,
+                    ApplicationContactType.HAKIJA,
+                    null,
+                    UUID.randomUUID()
+                )
+
+            put("$baseUrl/$id", request)
+                .andExpect(MockMvcResultMatchers.status().isBadRequest)
+                .andExpect(MockMvcResultMatchers.jsonPath("errorCode").value("HAI2010"))
+
+            verify {
+                applicationAuthorizer.authorizeApplicationId(
+                    id,
+                    PermissionCode.EDIT_APPLICATIONS.name
+                )
+                hakemusService.updateHakemus(id, request, USERNAME)
+            }
+        }
+
+        @Test
+        fun `returns 400 when request contain invalid contact`() {
+            val request =
+                HakemusUpdateRequestFactory.createFilledJohtoselvityshakemusUpdateRequest()
+            every {
+                applicationAuthorizer.authorizeApplicationId(
+                    id,
+                    PermissionCode.EDIT_APPLICATIONS.name
+                )
+            } returns true
+            every { hakemusService.updateHakemus(id, request, USERNAME) } throws
+                InvalidHakemusyhteyshenkiloException("Invalid contact")
+
+            put("$baseUrl/$id", request)
+                .andExpect(MockMvcResultMatchers.status().isBadRequest)
+                .andExpect(MockMvcResultMatchers.jsonPath("errorCode").value("HAI2011"))
+
+            verify {
+                applicationAuthorizer.authorizeApplicationId(
+                    id,
+                    PermissionCode.EDIT_APPLICATIONS.name
+                )
+                hakemusService.updateHakemus(id, request, USERNAME)
+            }
+        }
+
+        @Test
+        fun `returns application when it exists`() {
+            val expectedResponse = HakemusResponseFactory.create()
+            val request = expectedResponse.toUpdateRequest()
+
+            every {
+                applicationAuthorizer.authorizeApplicationId(
+                    id,
+                    PermissionCode.EDIT_APPLICATIONS.name
+                )
+            } returns true
+            every { hakemusService.updateHakemus(id, request, USERNAME) } returns expectedResponse
+
+            val response: HakemusResponse =
+                put("$baseUrl/$id", request)
+                    .andExpect(MockMvcResultMatchers.status().isOk)
+                    .andReturnBody()
+
+            assertThat(response).isEqualTo(expectedResponse)
```

Comment:
JSONAssert can be used to avoid needing deserializing on the response.

```suggestion
            val response =
                put("$baseUrl/$id", request)
                    .andExpect(MockMvcResultMatchers.status().isOk)
                    .andReturn()
                    .response
                    .getContentAsString(StandardCharsets.UTF_8)

            JSONAssert.assertEquals(
                expectedResponse.toJsonString(),
                response,
                JSONCompareMode.NON_EXTENSIBLE
            )
```

Reply:
Good to know, thanks!

Reply:
Had to add `.getContentAsString(Charset.forName("UTF-8"))` also because MockMvc uses ISO-8859-1 as default.